uaclient/tests/test_cli_attach_premium.py

```diff
@@ -24,9 +28,118 @@ def test_non_root_users_are_rejected(getuid):
+    )
+    @mock.patch("uaclient.clouds.identity.cloud_instance_factory")
+    @mock.patch("uaclient.clouds.identity.get_cloud_type", return_value="aws")
+    def test_aws_cloud_type_non_premium_returns_no_token(
+        self,
+        _get_cloud_type,
+        cloud_instance_factory,
+        request_premium_aws_contract_token,
+        capsys,
+    ):
+        """AWS clouds on non-premium images not return a token."""
+
+        def fake_instance_factory():
+            m_instance = mock.Mock()
+            m_instance.identity_doc = "pkcs7-validated-by-backend"
+            return m_instance
+
+        cloud_instance_factory.side_effect = fake_instance_factory
+        request_premium_aws_contract_token.side_effect = ContractAPIError(
+            util.UrlError(
+                "Server error", code=500, url="http://me", headers={}
+            ),
+            error_response={"message": "missing instance information"},
+        )
+        assert None is _get_contract_token_from_cloud_identity(FakeConfig())
+        out = capsys.readouterr()[0]
+        assert status.MESSAGE_UNSUPPORTED_PREMIUM == out.strip()
+
+    @mock.patch(
+        M_PATH + "contract.UAContractClient.request_premium_aws_contract_token"
+    )
+    @mock.patch("uaclient.clouds.identity.cloud_instance_factory")
+    @mock.patch("uaclient.clouds.identity.get_cloud_type", return_value="aws")
+    def test_raise_unexpected_errors(
+        self,
+        _get_cloud_type,
+        cloud_instance_factory,
+        request_premium_aws_contract_token,
+        capsys,
+    ):
+        """Any unexpected errors will be raised."""
+
+        def fake_instance_factory():
+            m_instance = mock.Mock()
+            m_instance.identity_doc = "pkcs7-validated-by-backend"
+            return m_instance
+
+        cloud_instance_factory.side_effect = fake_instance_factory
+        unexpected_error = ContractAPIError(
+            util.UrlError(
+                "Server error", code=500, url="http://me", headers={}
+            ),
+            error_response={"message": "something unexpected"},
+        )
+        request_premium_aws_contract_token.side_effect = unexpected_error
+
+        with pytest.raises(ContractAPIError) as excinfo:
+            _get_contract_token_from_cloud_identity(FakeConfig())
+        assert unexpected_error == excinfo.value
+
+    @mock.patch(
+        M_PATH + "contract.UAContractClient.request_premium_aws_contract_token"
+    )
+    @mock.patch("uaclient.clouds.identity.cloud_instance_factory")
+    @mock.patch("uaclient.clouds.identity.get_cloud_type", return_value="aws")
+    def test_return_token_from_contract_server_using_identity_doc(
+        self,
+        _get_cloud_type,
+        cloud_instance_factory,
+        request_aws_contract_token,
+        capsys,
+    ):
+        """Return token from the contract server using the identity."""
+
+        def fake_instance_factory():
```

Comment:
This identical definition is repeated several times, should it move to a top-level function that all the tests can share?

Reply:
agreed.